File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/RelatedActionsCategorizer.java
package cz.fb.manaus.reactor.categorizer;

import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.categorizer.SettledBetCategorizer;
import cz.fb.manaus.core.dao.BetActionDao;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.reactor.betting.action.BetUtils;
import cz.fb.manaus.spring.DatabaseComponent;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.HashSet;
import java.util.List;
import java.util.OptionalLong;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import static java.util.Optional.of;

@DatabaseComponent
public class RelatedActionsCategorizer implements SettledBetCategorizer {

    private static final Logger log = Logger.getLogger(RelatedActionsCategorizer.class.getSimpleName());


    @Autowired
    private BetActionDao betActionDao;
    @Autowired
    private BetUtils betUtils;
    @Autowired
    private List<RelatedActionsAwareCategorizer> relatedActionsAwareCategorizers;

    @Override
    public boolean isSimulationSupported() {
        return false;
    }

    @Override
    public Set<String> getCategories(SettledBet settledBet, BetCoverage coverage) {
        var market = settledBet.getBetAction().getMarket();
        var betActions = betActionDao.getBetActions(market.getId(),
                OptionalLong.of(settledBet.getSelectionId()), of(settledBet.getPrice().getSide()));
        if (betActions.isEmpty()) {
            log.log(Level.WARNING, "missing  bet actions ''{0}''", settledBet);
            return Set.of();
        }
        var current = betUtils.getCurrentActions(betActions);
        var result = new HashSet<String>();
        for (var categorizer : relatedActionsAwareCategorizers) {
            var partial = categorizer.getCategories(current, market);
            if (partial != null) {
                result.addAll(partial);
            }
        }
        return result;
    }
}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/categorizer/RelatedActionsCategorizer.kt
package cz.fb.manaus.reactor.categorizer

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.categorizer.RealizedBetCategorizer
import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.repository.BetActionRepository
import cz.fb.manaus.reactor.betting.action.BetUtils
import cz.fb.manaus.spring.ManausProfiles
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component
import java.util.logging.Level
import java.util.logging.Logger

@Component
@Profile(ManausProfiles.DB)
class RelatedActionsCategorizer(
        private val betActionRepository: BetActionRepository,
        private val relatedActionsAwareCategorizers: List<RelatedActionsAwareCategorizer>) : RealizedBetCategorizer {


    override val isSimulationSupported: Boolean = false

    override fun getCategories(realizedBet: RealizedBet, coverage: BetCoverage): Set<String> {
        val market = realizedBet.market
        val side = realizedBet.settledBet.price.side
        val selectionId = realizedBet.settledBet.selectionId
        // TODO filter in DB
        val betActions = betActionRepository.find(market.id)
                .filter { it.selectionId == selectionId && side == it.price.side }
        if (betActions.isEmpty()) {
            log.log(Level.WARNING, "missing  bet actions ''{0}''", realizedBet)
            return emptySet()
        }
        val current = BetUtils.getCurrentActions(betActions)
        val result = mutableSetOf<String>()
        for (categorizer in relatedActionsAwareCategorizers) {
            val partial = categorizer.getCategories(current, market)
            result.addAll(partial)
        }
        return result
    }

    companion object {
        private val log = Logger.getLogger(RelatedActionsCategorizer::class.java.simpleName)
    }
}





